src/cargo/core/resolver/mod.rs

```diff
@@ -427,10 +472,33 @@ impl Ord for DepsFrame {
 struct BacktrackFrame<'a> {
     context_backup: Context<'a>,
     deps_backup: BinaryHeap<DepsFrame>,
-    remaining_candidates: RcVecIter<Candidate>,
-    parent: Rc<Summary>,
+    remaining_candidates: RemainingCandidates,
+    parent: Summary,
     dep: Dependency,
-    features: Vec<String>,
+    features: Rc<Vec<String>>,
+}
+
+#[derive(Clone)]
+struct RemainingCandidates {
+    remaining: RcVecIter<Candidate>,
+}
+
+impl RemainingCandidates {
+    fn next(&mut self, prev_active: &[Summary]) -> Option<Candidate> {
+        // Filter the set of candidates based on the previously activated
+        // versions for this dependency. We can actually use a version if it
+        // precisely matches an activated version or if it is otherwise
+        // incompatible with all other activated versions. Note that we
+        // define "compatible" here in terms of the semver sense where if
+        // the left-most nonzero digit is the same they're considered
+        // compatible.
+        self.remaining.by_ref().map(|p| p.1).filter(|b| {
+            prev_active.iter().any(|a| *a == b.summary) ||
+                prev_active.iter().all(|a| {
+                    !compatible(a.version(), b.summary.version())
+                })
```

Comment:
Ah yeah this look actually doesn't matter too much, 100% of slowdown so far has stemmed from too much allocation, nothing's been "you should actually run less code" so far!